encoding/proto/proto_v2.go

```diff
@@ -0,0 +1,72 @@
+/*
+ *
+ * Copyright 2018 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package proto
+
+import (
+	"fmt"
+
+	"google.golang.org/grpc/encoding"
+	"google.golang.org/grpc/mem"
+	"google.golang.org/protobuf/proto"
+)
+
+func init() {
+	encoding.RegisterCodecV2(&codecV2{})
+}
+
+// codec is a CodecV2 implementation with protobuf. It is the default codec for
+// gRPC.
+type codecV2 struct{}
+
+var _ encoding.CodecV2 = (*codecV2)(nil)
+
+func (c *codecV2) Marshal(v any) (mem.BufferSlice, error) {
+	vv := messageV2Of(v)
+	if vv == nil {
+		return nil, fmt.Errorf("proto: failed to marshal, message is %T, want proto.Message", v)
+	}
+
+	pool := mem.DefaultBufferPool()
```

Comment:
Yeah that definitely feels clunky. Technically however, it's still possible when it's done the other way around. When creating the channel, the caller can pass in the pool with `WithBufferPool` and a `CodecV2` implementation that uses the same pool. It's equally clunky but at least leaves the door open to that. The idea here is that the default pool should be good enough for 99% of use cases, and extreme power users will be fine jumping through hoops for their performance gains from custom pools. Thoughts?

Reply:
Remember that the default marshaller outputs everything to a single large buffer instead of smaller chunked buffers that play better with buffer pools. I think anyone that requires their custom pools will likely also provide custom codecs.